File: go/supersandbox/terminal.go
package supersandbox

import (
	"context"
	"fmt"
	"strings"

	"github.com/gorilla/websocket"
)

// TerminalSession is a WebSocket PTY connection to a running sandbox.
type TerminalSession struct {
	conn *websocket.Conn
}

// TerminalOption configures a terminal session.
type TerminalOption func(*terminalParams)

type terminalParams struct {
	cols int
	rows int
}

// WithSize sets the initial terminal size.
func WithSize(cols, rows int) TerminalOption {
	return func(p *terminalParams) {
		p.cols = cols
		p.rows = rows
	}
}

func newTerminalSession(ctx context.Context, h *httpClient, sandboxID string, opts ...TerminalOption) (*TerminalSession, error) {
	p := &terminalParams{cols: 80, rows: 24}
	for _, o := range opts {
		o(p)
	}

	wsURL := strings.Replace(h.baseURL, "https://", "wss://", 1)
	wsURL = strings.Replace(wsURL, "http://", "ws://", 1)
	wsURL = fmt.Sprintf("%s/sandboxes/%s/terminal?cols=%d&rows=%d", wsURL, sandboxID, p.cols, p.rows)

	dialer := websocket.Dialer{}
	headers := map[string][]string{
		"OPEN-SANDBOX-API-KEY": {h.apiKey},
	}
	conn, _, err := dialer.DialContext(ctx, wsURL, headers)
	if err != nil {
		return nil, fmt.Errorf("terminal dial: %w", err)
	}
	return &TerminalSession{conn: conn}, nil
}

// Send writes input to the terminal (keyboard input).
func (t *TerminalSession) Send(data string) error {
	return t.conn.WriteMessage(websocket.TextMessage, []byte(data))
}

// Receive reads one message from the terminal output.
func (t *TerminalSession) Receive() (string, error) {
	_, msg, err := t.conn.ReadMessage()
	if err != nil {
		return "", err
	}
	return string(msg), nil
}

// Stream returns a channel that emits terminal output lines until the context
// is cancelled or the connection closes.
func (t *TerminalSession) Stream(ctx context.Context) <-chan StreamLine {
	ch := make(chan StreamLine, 64)
	go func() {
		defer close(ch)
		for {
			select {
			case <-ctx.Done():
				return
			default:
			}
			_, msg, err := t.conn.ReadMessage()
			if err != nil {
				if ctx.Err() == nil {
					ch <- StreamLine{Err: err}
				}
				return
			}
			ch <- StreamLine{Text: string(msg)}
		}
	}()
	return ch
}

// Close shuts down the terminal session.
func (t *TerminalSession) Close() error {
	return t.conn.Close()
}

File: python/src/supersandbox/terminal.py
"""WebSocket terminal session for a running sandbox."""

from __future__ import annotations

import asyncio
from typing import AsyncIterator, Optional
from urllib.parse import urlparse

import websockets
from websockets.exceptions import ConnectionClosed

from .exceptions import TerminalError


class TerminalSession:
    """Interactive WebSocket PTY session.

    Obtain via ``async with sandbox.terminal() as term:``.

    Usage::

        async with sandbox.terminal() as term:
            await term.send("ls /workspace\\n")

            # Single chunk (returns "" on timeout)
            output = await term.receive(timeout=2.0)

            # Stream all output until connection closes
            async for chunk in term.stream():
                print(chunk, end="", flush=True)
    """

    def __init__(self, ws_url: str, api_key: str) -> None:
        self._ws_url = ws_url
        self._api_key = api_key
        self._ws: Optional[websockets.WebSocketClientProtocol] = None

    async def __aenter__(self) -> "TerminalSession":
        try:
            self._ws = await websockets.connect(
                self._ws_url,
                additional_headers={"Open-Sandbox-Api-Key": self._api_key},
                ping_interval=20,
                ping_timeout=10,
            )
        except Exception as exc:
            raise TerminalError(f"Failed to connect to terminal: {exc}") from exc
        return self

    async def __aexit__(self, *_: object) -> None:
        if self._ws:
            await self._ws.close()
            self._ws = None

    async def send(self, text: str) -> None:
        """Send keystrokes or a command to the terminal."""
        if not self._ws:
            raise TerminalError("Terminal session is not open.")
        await self._ws.send(text)

    async def receive(self, timeout: float = 5.0) -> str:
        """Receive a single chunk of terminal output.

        Returns an empty string if no output arrives within ``timeout`` seconds.
        """
        if not self._ws:
            raise TerminalError("Terminal session is not open.")
        try:
            return await asyncio.wait_for(self._ws.recv(), timeout=timeout)  # type: ignore[return-value]
        except asyncio.TimeoutError:
            return ""

    async def stream(self) -> AsyncIterator[str]:
        """Yield terminal output chunks until the connection closes."""
        if not self._ws:
            raise TerminalError("Terminal session is not open.")
        try:
            async for message in self._ws:
                yield message  # type: ignore[misc]
        except ConnectionClosed:
            return
